### PR TITLE
Enhanced local variable table

### PR DESCRIPTION
**Pull request by smowton**
This PR replaces the existing simple local variable table code, which introduces a distinct variable per local variable table entry, with a simple control-flow analysis that merges table entries that flow into one another.

Recommend looking at each commit individually, as they introduce new functionality, use it, then factor into separate files.

### CODE REVIEW

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 1]
```diff
@@ -505,6 +545,420 @@ void java_bytecode_convert_methodt::convert(
...
+  if(to!=from)
+    var.holes.push_back({from,to-from});
+}
+
+void java_bytecode_convert_methodt::find_initialisers_for_slot(
+  local_variable_table_with_holest::iterator firstvar,
+  local_variable_table_with_holest::iterator varlimit,
+  const address_mapt& amap,
+  const java_cfg_dominatorst& doms)
+{
```
Comment (tautschnig) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: This function is very long, would it be possible to factor out parts (e.g., loop bodies) into separate methods?

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 2]
```diff
@@ -2492,13 +2495,13 @@ codet java_bytecode_convert_methodt::convert_instructions(
...
       if(v.symbol_expr.get_identifier()==irep_idt())
         continue;
-      code_declt d(v.symbol_expr);
       auto &block=get_block_for_pcrange(
         root,
         root_block,
         v.start_pc,
         v.start_pc+v.length,
         std::numeric_limits<unsigned>::max());
+      code_declt d(v.symbol_expr);
```
Comment (tautschnig) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: What is the need for this line shuffling?

File: src/java_bytecode/java_local_variable_table.cpp [hunk 3]
```diff
@@ -5,6 +5,7 @@
 #include "java_types.h"
 
 #include <climits>
+#include <iostream>
```
Comment (tautschnig) [inline src/java_bytecode/java_local_variable_table.cpp]: Why do you need iostream?

File: src/java_bytecode/java_local_variable_table.cpp [hunk 4]
```diff
@@ -522,7 +522,7 @@ java_bytecode_convert_methodt::find_variable_for_slot(
     {
       bool found_hole=false;
       for(auto &hole : var.holes)
-        if(address>=hole.start_pc && address<(hole.start_pc-hole.length))
+        if(address>=hole.start_pc && address<(hole.start_pc+hole.length))
```
Comment (tautschnig) [inline src/java_bytecode/java_local_variable_table.cpp]: So where is the change to the fixed test?

**Comment by smowton** [comment]
It's a real pain to merge the hunks in to the right commit-- do you know of a good way to do that automatically? (e.g. find the last commit to touch this line and squash the relevant hunk into it)?

**Comment by tautschnig** [comment]
No, I'm afraid I don't know of an automated way of doing it - I generally end up with running git blame in one window and git add -p in the other. There's probably a much better use of your time than massaging those commits. Leave them as they are, I'd say.

**Comment by smowton** [comment]
All pre-requisites are in and rebased upon, this should now be mergeable.

**Pull request merged**